defaultConfig.txt
#This is a comment

##Domain
domain=beta.risistar.fr
userAgent=Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:67.0) Gecko/20100101 Firefox/67.0

## Features activation
activateAutoBuild=True
activateCustomBuildOrders=False
activateAutoResearch=True
activateAutoFleetScan=True
activateDefenderDiscordPing=False
activateAutoEvasion=False
activateAutoExpedition=False
activateAutoColonization=False
activatePickingOfficers=False
activateAllianceScan=False

## AutoBuild Settings
# Ratio that will determine the level of your robot factory based on the level of your metal mine.
# Round to the first inferior integer
# Ex: metal mine 17 -> 17/3 = 5.66 -> robot factory will be upgraded to level 5.
robotRatio=3
# Determine when the first level of robot factory should be done based on the level of your metal mine
# Set it to 6 means that the robot factory will only be planned when your metal mine is level 6 or higher
robotStartingLevel=0

##Custom build order settings
customBuildOrdersDirectoryName=buildOrders
customBuildOrdersPairingFile=customBuildOrdersPairingFile.txt

##AutoFleetScan settings
minimumTimeBetweenScans=30
randomAdditionnalTimeBetweenScans=10

## Auto Tech settings
techsPickingOrderFile=techsPickingOrder.txt

##Ennemy fleet discord ping
webhookUrl=None
idToPing=None
#the message to be sent
#it can use targetPlanet.position or .name or fleet.ttd
customAttackNotificationMessage=Attack on {targetPlanet.name} ({targetPlanet.position}) in {fleet.ttd}

##AutoEvasion settings
#This indicates where your fleet will travel to when evading an attack
escapeTarget=1:50:5:1
minimumSpottingTime=300

##AutoExpedition settings
#This indicates if only one expedition is launched at a time (one per fleet scan)
#If false, it will launch all missing expedition directly (which is impossible for a human being)
launchExpeditionSeparately=True

##Getting Officers settings
officersPickingOrderFile=officersPickingOrder.txt

##Alliance scan
#Put your url, but don't disclose it to anyone other than your alliance member
shrib=AAAAAAAAAA
shribCookie=AAAAAAAAAA

## Watchdog
# The default delay in seconds between each reinitialization of the AI
watchdogDelay=1800
# If an exception is raised, the watchdog will wake sooner
# This indicates the amount in seconds to subtract from the remaining time until the watchdog wake up
watchdogExceptionDelay=60
# After executing a task in the watchdogEarlyDelay seconds before the watchdog wakes up,
# If there is no mid to high priority task scheduled in the next watchdogWakeDuration seconds,
# Then wake the watchdog early
watchdogEarlyDelay=60
# watchdogWakeDuration should be a bit more than the time to initialize the AI
watchdogWakeDuration=5


Config.py
import os.path

class Config:
    def __init__(self):
        self.domain = None
        self.userAgent = None
        self.activateAutoBuild = None
        self.activateCustomBuildOrders = None
        self.activateAutoResearch = None
        self.activateAutoFleetScan = None
        self.activateDefenderDiscordPing = None
        self.activateAutoEvasion = None
        self.activateAutoExpedition = None
        self.activateAutoColonization = None
        self.activatePickingOfficers = None
        self.activateAllianceScan = None
        self.robotRatio = None
        self.robotStartingLevel = None
        self.customBuildOrdersDirectoryName = None
        self.customBuildOrdersPairingFile = None
        self.techsPickingOrderFile = None
        self.minimumTimeBetweenScans = None
        self.randomAdditionnalTimeBetweenScans = None
        self.webhookUrl = None
        self.idToPing = None
        self.customAttackNotificationMessage = None
        self.escapeTarget = None
        self.minimumSpottingTime = None
        self.launchExpeditionSeparately = None
        self.officersPickingOrderFile = None
        self.shrib = None
        self.shribCookie = None
        self.watchdogDelay = None
        self.watchdogExceptionDelay = None
        self.watchdogEarlyDelay = None
        self.watchdogWakeDuration = None

    @staticmethod
    def load():
        config = Config()
        config.loadFromFile("defaultConfig.txt")
        if os.path.isfile("config.txt"):
            config.loadFromFile("config.txt")
        return config

    def loadFromFile(self, filePath):
        with open(filePath) as file:
            l = file.readline()
            while l != "":
                l = l.replace("\n", "")
                if len(l) and l[0] != "#":
                    s = l.split("=")
                    key = s[0].replace(" ", "")
                    value = s[1]
                    if key == "domain":
                        self.domain = value
                    elif key == "userAgent":
                        self.userAgent = value
                    elif key == "activateAutoBuild":
                        self.activateAutoBuild = value == "True"
                    elif key == "activateCustomBuildOrders":
                        self.activateCustomBuildOrders = value == "True"
                    elif key == "activateAutoResearch":
                        self.activateAutoResearch = value == "True"
                    elif key == "activateAutoFleetScan":
                        self.activateAutoFleetScan = value == "True"
                    elif key == "activateDefenderDiscordPing":
                        self.activateDefenderDiscordPing = value == "True"
                    elif key == "activateAutoEvasion":
                        self.activateAutoEvasion = value == "True"
                    elif key == "activateAutoExpedition":
                        self.activateAutoExpedition = value == "True"
                    elif key == "activateAutoColonization":
                        self.activateAutoColonization = value == "True"
                    elif key == "activatePickingOfficers":
                        self.activatePickingOfficers = value == "True"
                    elif key == "activateAllianceScan":
                        self.activateAllianceScan = value == "True"
                    elif key == "robotRatio":
                        self.robotRatio = int(value)
                    elif key == "robotStartingLevel":
                        self.robotStartingLevel = int(value)
                    elif key == "customBuildOrdersDirectoryName":
                        self.customBuildOrdersDirectoryName = value
                    elif key == "customBuildOrdersPairingFile":
                        self.customBuildOrdersPairingFile = value
                    elif key == "techsPickingOrderFile":
                        self.techsPickingOrderFile = value
                    elif key == "minimumTimeBetweenScans":
                        self.minimumTimeBetweenScans = int(value)
                    elif key == "randomAdditionnalTimeBetweenScans":
                        self.randomAdditionnalTimeBetweenScans = int(value)
                    elif key == "webhookUrl":
                        self.webhookUrl = value
                    elif key == "idToPing":
                        self.idToPing = value
                    elif key == "customAttackNotificationMessage":
                        self.customAttackNotificationMessage = value
                    elif key == "escapeTarget":
                        self.escapeTarget = [int(x) for x in value.split(":")]
                    elif key == "minimumSpottingTime":
                        self.minimumSpottingTime = int(value)
                    elif key == "officersPickingOrderFile":
                        self.officersPickingOrderFile = value
                    elif key == "launchExpeditionSeparately":
                        self.launchExpeditionSeparately = value == "True"
                    elif key == "shrib":
                        self.shrib = value
                    elif key == "shribCookie":
                        self.shribCookie = value
                    elif key == "watchdogDelay":
                        self.watchdogDelay = int(value)
                    elif key == "watchdogExceptionDelay":
                        self.watchdogExceptionDelay = int(value)
                    elif key == "watchdogEarlyDelay":
                        self.watchdogEarlyDelay = int(value)
                    elif key == "watchdogWakeDuration":
                        self.watchdogWakeDuration = int(value)
                l = file.readline()

    def getError(self):
        if None in self.__dict__.values():
            return "Some settings aren't set !"
        if self.activateCustomBuildOrders and not self.activateAutoBuild:
            return "Custom build orders is activated but not auto build !"
        if self.activateCustomBuildOrders and self.customBuildOrdersDirectoryName is None:
            return "Custom build orders is activated but the build orders directory isn't specified !"
        if self.activateCustomBuildOrders and self.customBuildOrdersPairingFile is None:
            return "Custom build orders is activated but the build orders pairing file isn't specified !"
        if self.activateAutoEvasion and not self.activateAutoFleetScan:
            return "Auto evasion is activated but not auto fleet scan !"
        if self.activateAutoExpedition and not self.activateAutoFleetScan:
            return "Auto expedition is activated but not auto fleet scan !"
        if self.activateDefenderDiscordPing and not self.activateAutoFleetScan:
            return "Defender ping is activated but not auto fleet scan !"
        if self.activateDefenderDiscordPing and self.webhookUrl == "None":
            return "Defender ping is activated but the webhook url isn't configurated !"
        if self.activateDefenderDiscordPing and self.idToPing == "None":
            return "Defender ping is activated but the user id to ping isn't configurated !"
        if self.activatePickingOfficers and self.officersPickingOrderFile is None:
            return "Officer picking is activated but the order isn't configurated !"
        if self.activateAllianceScan and (self.shrib is None or self.shribCookie is None):
            return "You need to have a shrib and a cookie !"
        if self.activateAllianceScan and not self.activateAutoFleetScan:
            return "Alliance scan is activated but not auto fleet scan !"
        if self.watchdogDelay is None:
            return "watchdogDelay isn't defined"
        if self.watchdogDelay <= 0:
            return "watchdogDelay should be strictly positive (for example 1800 ie 30 minutes)"
        if self.watchdogExceptionDelay is None:
            return "watchdogExceptionDelay isn't defined (in doubt, put it at 0)"
        if self.watchdogEarlyDelay is None:
            return "watchdogEarlyDelay isn't defined (in doubt, put it at 0)"
        if self.watchdogWakeDuration is None:
            return "watchdogWakeDuration isn't defined (in doubt, put it at 5)"
        return None
